stable_plugins/visualization/file_types/qasm_visualization.py

```diff
@@ -0,0 +1,252 @@
+        schema = QasmInputParametersSchema()
+        return Response(
+            render_template(
+                "qasm_visualization.html",
+                name=plugin.name,
+                version=plugin.version,
+                schema=schema,
+                valid=valid,
+                values=data,
+                errors=errors,
+                example_values=url_for(f"{QASM_BLP.name}.MicroFrontend"),
+            )
+        )
+
+
+@QASM_BLP.route("/circuits/")
+def get_circuit_image():
+    circuit_url = request.args.get("circuit-url", None)
+    if circuit_url is None:
+        abort(HTTPStatus.NOT_FOUND)
+    filename = hashlib.md5(circuit_url.encode("utf-8")).hexdigest() + ".png"
+    path = pathlib.Path(__file__).parent.absolute() / "img" / filename
+    if os.path.exists(path):
+        return send_file(path, mimetype="image/png")
+    try:
+        with open_url(circuit_url) as quasm_response:
+            circuit_qasm = quasm_response.text
+    except HTTPError:
+        TASK_LOGGER.info(f"Could not load circuit from {circuit_url}")
+        abort(HTTPStatus.NOT_FOUND)
+    circuit = QuantumCircuit.from_qasm_str(circuit_qasm)
+    fig = circuit.draw(output="mpl")
+    fig.savefig(path)
+    return send_file(path, mimetype="image/png")
+
+
+@QASM_BLP.route("/process/")
+class ProcessView(
+    MethodView
+):  # FIXME decide on a somewhat useful implementation for this (or remove completely!)
+    """Start a long running processing task."""
+
+    @QASM_BLP.arguments(QasmInputParametersSchema(unknown=EXCLUDE), location="form")
+    @QASM_BLP.response(HTTPStatus.SEE_OTHER)
+    @QASM_BLP.require_jwt("jwt", optional=True)
+    def post(self, arguments):
+        """Start the demo task."""
+        db_task = ProcessingTask(task_name=demo_task.name, parameters=dumps(arguments))
+        db_task.save(commit=True)
+
+        # all tasks need to know about db id to load the db entry
+        task: chain = demo_task.s(db_id=db_task.id) | save_task_result.s(db_id=db_task.id)
+        # save errors to db
+        task.link_error(save_task_error.s(db_id=db_task.id))
+        task.apply_async()
+
+        db_task.save(commit=True)
+
+        return redirect(
+            url_for("tasks-api.TaskView", task_id=str(db_task.id)), HTTPStatus.SEE_OTHER
+        )
+
+
+class QasmVisualization(QHAnaPluginBase):
+    name = _plugin_name
+    version = __version__
+    description = "Visualizes QASM data."
+    tags = ["visualization"]
+
+    def __init__(self, app: Optional[Flask]) -> None:
+        super().__init__(app)
+
+        # create folder for circuit images
+        pathlib.Path(__file__).parent.absolute().joinpath("img").mkdir(
+            parents=True, exist_ok=True
+        )
+
+    def get_api_blueprint(self):
+        return QASM_BLP
+
+
+TASK_LOGGER = get_task_logger(__name__)
+
+
+@CELERY.task(name=f"{QasmVisualization.instance.identifier}.demo_task", bind=True)
+def demo_task(self, db_id: int) -> str:
```

Comment:
If you want to do the work (which would not be too much in your case) this task could also calculate the circuit visualization and then output it like any other processing plugin task would to the file store. This way, the plugin could also be used as a processing plugin. (This is harder to do for plugins with only client side visualization, which is why the existing plugins do not do this yet.)

Reply:
The `qasm-visualization` plugin can now be used as processing and visualization plugin. Images are stored as files (like the outputs of other processing plugins) and a mapping between `url_hash`es ans file paths is stored in the PluginState. Images are created through a celery task. If an image is not yet created, flask responds with status code `503 Service Unavailable` and the HTTP header [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) set. Since the `Retry-After` header doesn't work for all browsers, the "timeout before retry" is implemented in [`qasm-visualization.html`](https://github.com/UST-QuAntiL/qhana-plugin-runner/blob/94d112c27f1f1f083334afa0fbdc7d0011e2a5bb/stable_plugins/visualization/file_types/qasm_visualization_templates/qasm_visualization.html#L106-L124).